### PR TITLE
Feature/report ec2cost

### LINKED ISSUE #21: add ability to pull ec2 cost data

### PR DESCRIPTION
**Pull request by bspindler**
Closes #21

### CODE REVIEW

File: ts/commands/ReportCommands.ts [hunk 1]
```diff
@@ -19,6 +19,22 @@ class ReportCommands {
...
         reportService.list(config, logger);
       });
+
+    (caporal as any)
+      .command('report ec2cost', 'Retrieve ec2 cost data grouped and aggregated.')
+      .option('--username', 'Metricly Username')
+      .option('--password', 'Metricly Password')
+      .option('--profile', 'Metricly profile', /.*/, 'default')
+      .option('--format', 'format options: text, json', ['text', 'json'], 'text')
+      .option('--timerange', 'relative timerange options: latestDay, monthToDate, lastMonth', ['latestDay', 'monthToDate', 'lastMonth'], 'latestDay')
```
Comment (bspindler) [inline ts/commands/ReportCommands.ts]: based on some other reports, to be consistent I think I should rename to `period`

File: ts/model/report/ReportContent.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { ViewDefinition } from './ViewDefinition';
+
+export class ReportContent {
+    public static fromJSON(json: string): ReportContent {
+        const user = Object.create(ReportContent.prototype);
```
Comment (TheConnMan) [inline ts/model/report/ReportContent.ts]: Why is this variable named `user`?

File: ts/model/report/ReportScope.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+export class ReportScope {
+    public view: string;
+    public parameters: {};
+    public rowLimit: number;
+    public withViewDefinitions: boolean;
+
+    constructor(view, parameters = {}, rowLimit = 20, withViewDefinitions = false) {
```
Comment (TheConnMan) [inline ts/model/report/ReportScope.ts]: If you add `public` in front of these variables you don't need to duplicate their definition in lines 2-5, nor will you need to assign them in lines 8-11. Take a look at https://www.typescriptlang.org/docs/handbook/classes.html#static-properties.

This should reduce the class down to the following:

```typescript
export class ReportScope {
    constructor(public view: any, public parameters = {}, public rowLimit = 20, public withViewDefinitions = false) { }
}
```

File: ts/services/ReportService.ts [hunk 4]
```diff
@@ -1,36 +1,150 @@
...
+
+            // json
+            if (config.format === 'json') {
+              logger.info(JSON.stringify(groupedCost, null, 2));
+            }
+          }
+        }
+      }
+    } catch (e) {
+      logger.error('There was an error getting ec2 cost data: ' + e);
```
Comment (TheConnMan) [inline ts/services/ReportService.ts]: EC2 should be capitalized.

File: ts/services/ReportService.ts [hunk 5]
```diff
@@ -1,36 +1,150 @@
...
+      });
+      return response;
+    } catch (e) {
+      logger.error('There was an error fetching the report content: ' + e);
+    }
+  }
+
+  private getStartDate(endDate: Date, timerange: string) {
+
+    return timerange === 'lastMonth' ?
```
Comment (TheConnMan) [inline ts/services/ReportService.ts]: This is a very long (double) ternary operation. Can you break it up into something more readable? E.g.

```typescript
if (timerange === 'lastMonth') {
    return moment(endDate).utc().subtract(1, 'months').startOf('month').toISOString();
}
if (timerange === 'monthToDate') {
    return moment(endDate).utc().startOf('month').toISOString();
}
return moment(endDate).toISOString();
```

Or use a switch statement, which this seems like it is.

File: ts/services/ReportService.ts [hunk 6]
```diff
@@ -1,36 +1,150 @@
...
+  private async getElementsInScope(config, logger, report: Report): Promise<ReportContent> {
+
+    const reportScope = new ReportScope('elementsInScope', {
+      elementTypes: ['EC2', 'WINSVR', 'SERVER'],
+      endDate: this.getEndDate(report.endDate, config.timerange),
+      startDate: this.getStartDate(report.endDate, config.timerange)
+    },
+    config.rowlimit);
+    return this.doReportContentPost(config, logger, reportScope, report);
+
```
Comment (TheConnMan) [inline ts/services/ReportService.ts]: This space is unnecessary.

File: ts/services/ReportService.ts [hunk 7]
```diff
@@ -1,36 +1,150 @@
...
       if (config.format === 'text') {
         logger.info('The following reports are available:');
-        logger.info(response.reports.sort((rpt1, rpt2) => {
+        logger.info(reports.sort((rpt1, rpt2) => {
           return rpt1.name.localeCompare(rpt2.name);
-        }).map((rpt) => {
-          return rpt.name + ' (ID: ' + rpt.id + ')';
+        }).map((report) => {
+          return report.name + ' (Type: ' + report.type + ', ID: ' + report.id + ')';
```
Comment (TheConnMan) [inline ts/services/ReportService.ts]: Template literals are preferred over string concatenation: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals. E.g.

```typescript
return `${report.name} (Type: ${report.type}, ID: ${report.id})`;
```

File: ts/model/report/ReportContent.ts (hunk 2 above)
Comment (bspindler) [reply ts/model/report/ReportContent.ts]: mistake

File: ts/model/report/ReportScope.ts (hunk 3 above)
Comment (bspindler) [reply ts/model/report/ReportScope.ts]: brilliant, thanks for that

File: ts/services/ReportService.ts (hunk 5 above)
Comment (TheConnMan) [reply ts/services/ReportService.ts]: I prefer early returns over long, broken up lines. I don't have strong enough grounds for a switch if you like it, though, so I'm fine with you leaving it.

**Pull request merged**